src/Interpreters/GraceHashJoin.cpp

```diff
@@ -579,42 +606,57 @@ GraceHashJoin::InMemoryJoinPtr GraceHashJoin::makeInMemoryJoin()
     return std::make_unique<InMemoryJoin>(table_join, right_sample_block, any_take_last_row);
 }
 
+Block GraceHashJoin::prepareRightBlock(const Block & block)
+{
+    return HashJoin::prepareRightBlock(block, hash_join_sample_block);
+}
+
 void GraceHashJoin::addJoinedBlockImpl(Block block)
 {
     Buckets buckets_snapshot = getCurrentBuckets();
-    Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, block, buckets_snapshot.size());
     size_t bucket_index = current_bucket->idx;
+    Block current_block;
+
+    {
+        Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, block, buckets_snapshot.size());
+        flushBlocksToBuckets<JoinTableSide::Right>(blocks, buckets_snapshot, bucket_index);
+        current_block = std::move(blocks[bucket_index]);
+    }
 
     // Add block to the in-memory join
-    if (blocks[bucket_index].rows() > 0)
+    if (current_block.rows() > 0)
     {
         std::lock_guard lock(hash_join_mutex);
 
-        hash_join->addJoinedBlock(blocks[bucket_index], /* check_limits = */ false);
-        bool overflow = !fitsInMemory();
+        hash_join->addJoinedBlock(current_block, /* check_limits = */ false);
 
-        if (overflow)
-        {
-            auto right_blocks = hash_join->releaseJoinedBlocks();
-            right_blocks.pop_back();
+        if (!hasMemoryOverflow(hash_join))
+            return;
 
-            for (const auto & right_block : right_blocks)
-                blocks.push_back(right_block);
-        }
+        current_block = {};
+
+        auto right_blocks = hash_join->releaseJoinedBlocks(/* restructure */ false);
+
+        buckets_snapshot = rehashBuckets(buckets_snapshot.size() * 2);
 
-        while (overflow)
         {
-            buckets_snapshot = rehashBuckets(buckets_snapshot.size() * 2);
+            Blocks current_blocks;
+            current_blocks.reserve(right_blocks.size());
+            for (const auto & right_block : right_blocks)
+            {
+                Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, right_block, buckets_snapshot.size());
+                flushBlocksToBuckets<JoinTableSide::Right>(blocks, buckets_snapshot, bucket_index);
```

Comment:
Also scatter and flush one by one, because we already run out of memory and didn't want to keep them all plus scattered ones